src/H5FDcore.c

```diff
@@ -491,10 +458,30 @@ H5FD__init_package(void)
 hid_t
 H5FD_core_init(void)
 {
-    hid_t ret_value = H5I_INVALID_HID; /* Return value */
+    char *lock_env_var = NULL;            /* Environment variable pointer */
+    hid_t ret_value    = H5I_INVALID_HID; /* Return value */
 
     FUNC_ENTER_NOAPI(H5I_INVALID_HID)
 
+    /* It is possible that an application will call this routine through
+     * a `H5FD_*` symbol (`H5FD_FAMILY`, `H5FD_MULTI`, `H5FD_SEC2`, et
+     * cetera) before the library has had an opportunity to initialize.
+     * Call H5_init_library() to make sure that the library has been
+     * initialized before this VFD tries to initialize.
+     */
+    if (H5_init_library() < 0) {
+        HGOTO_ERROR(H5E_FUNC, H5E_CANTINIT, H5I_INVALID_HID, "library initialization failed")
+    }
```

Comment:
It's relatively annoying I think that we have to add that `H5_init_library()` here and there in the library... it would be good to find a way to get rid of that to make sure nobody eventually forgets one... it's not that clear to me why that's actually needed here

Reply:
An application can enter the library for the first time by evaluating one of the H5FD_<vfd> symbols such as H5FD_CORE. These H5_init_library() calls are there to make sure that the library is initialized before the VFD initialization proceeds.

Reply:
One thing that I could do is create a wrapper function, `hid_t H5FD_perform_init(hid_t (*init)(void))`, that does `H5_init_library`, checks its status, and on success calls `init`. Then each of the `H5FD_<vfd>` symbols can be `#define H5FD_<vfd> H5FD_perform_init(H5FD_<vfd>_init)`.

Right now some VFDs are effectively "internal" to the library and others are "external". The external ones cannot "see" H5_init_library and other library-internal functions, so they make sure the library is initialized with a call to H5open. H5FD_perform_init() can smooth over that difference.

Reply:
Yes anything that would make things easier would be good. One concern that I have for now is for external VFDs and VOLs maintainers, that init mechanism should be as transparent as possible for them.

Reply:
We could reduce a lot of complexity if we simply required people to explicitly initialize the library, as in MPI with MPI_Init().

Reply:
right or we could also just simply use constructor attributes since we don't use any args for init. Would also be cleaner than the atexit which would become a destructor.

Reply:
My most recent change introduces `hid_t H5FDperform_init(hid_t (*init)(void))` and uses it for the H5FD_<vfd> symbols.

Reply:
definitely looks nicer I think.